### ISSUE TITLE
SpringBoot 3 Declarative HTTP client interface, incompatibility with SpringBoot Resilience4J annotations

### ISSUE THREAD

**Issue by patpatpat123** [resilience4j/resilience4j#1848]
Resilience4j version: latest

Java version: 17

Hello Resiliene4J team,

I am a happy user of SpringBoot Resilience4J, especially circuit breaker, fallback and annotation.

For instance, my typical use looks something like this:

```
    @CircuitBreaker(name = "myCircuitBreaker", fallbackMethod = "fallback")
    public Mono<SomeResponse> sendPostToSomeAPI(SomeRequest someRequest) {
        return webClient.mutate().baseUrl("http://some-url").build().post().uri("/api").body(BodyInserters.fromValue(someRequest)).retrieve().bodyToMono(SomeResponse.class);
    }

public Mono<Object> fallBack(Exception exception) {
        //complexe failure handling
        return Mono.error(new ...);
    }
```

Which is working very well, very happy, I see the circuit breaker, most of all, I do see the fallback methods beings invoked appropriately.

Recently, with Spring 3.0, there is this new feature of declarative http client, very well demonstrated here: https://www.youtube.com/watch?v=3NcmlrumSOc

With this new approach, I managed to refactor my old WebFlux WebClient into this new nicer way of doing things:

```
    @PostExchange(url = "http://some-url/api")
    Mono<SomeResponse> sendPostToSomeAPI(@RequestBody SomeRequest someRequest);
```

Unfortunately, now it is an interface, I am having trouble with defining the fallback method:

```
    @PostExchange(url = "http://some-url/api")
    @CircuitBreaker(name = "myCircuitBreaker", fallbackMethod = "fallback") //this is not working, since the real implementation of the fallback method is not, and cannot be in this new declarative interface
    Mono<SomeResponse> sendPostToSomeAPI(@RequestBody SomeRequest someRequest);
```

Hence, I would like to raise this issue, stating a possible incompatibility between the fallback annotation and its "in the same class fallback method" and the new Spring 3.0 declarative http clients.

Thank you

**Comment by medimohammadise** [comment]
CircuiteBreaker and retry itself work for you and you only have problem with fallback method? Asking because for me it does not work at all with the new http interface

**Comment by RobWin** [comment]
Hi,
i didn't have time to look into it.
But I believe it requires a similar dedicated implementation like resilience4j-openfeign or resilience4j-retrofit.
It won't work out-of-the-box.
Any PR is very welcome.

**Comment by patpatpat123** [comment]
Hello @RobWin ,

Since this is my first message to you in 2025, I wanted to start by wishing you a happy new year and all the best 2025.

I was just wondering if there is any progress on this, this feature seems quite usefull.

**Comment by RobWin** [comment]
Hi,
No, i don't think that anyone is working on it.

**Comment by LivingLikeKrillin** [comment]
Hello @RobWin and others involved in this issue,

I'm curious about the current status of this feature. Has it been resolved or officially supported yet?

If it's not resolved yet, I'd like to suggest a possible approach:

My proposed solution:
Extend the existing @CircuitBreaker annotation by adding fallbackClass and fallbackMethod attributes, enabling clear definition of fallback logic even when using interface-based declarative HTTP clients.
At runtime, the fallback instance would be resolved via Spring ApplicationContext if present, or instantiated via reflection (cached for performance) in pure Java scenarios.
This approach preserves existing functionality and neatly integrates support for the declarative HTTP client scenario introduced in Spring Boot 3.
If this proposal sounds reasonable, I'd be happy to contribute a PR for this improvement.

Looking forward to your feedback, and thank you as always for your great work!

**Comment by RobWin** [comment]
Hi,
as far as I remember the annotations already support Spring SpEL to reference methods in other classes as a fallback.
is that what you need?

**Comment by patpatpat123** [comment]
Does anyone have an example with Spring SpEL and a fallback class in other classes?

**Comment by RobWin** [comment]
https://github.com/resilience4j/resilience4j/blob/master/resilience4j-spring6/src/test/java/io/github/resilience4j/spring6/CircuitBreakerDummyService.java#L117

But I dont remember if its possible to point to a different bean with SpEL.

**Comment by patpatpat123** [comment]
Hey @RobWin , I was aware of this link, and thanks for pasting it.

As of now, the fallback method seems to point to TestDummyService, with a default method.

"But I dont remember if its possible to point to a different bean with SpEL."
-> This is what I am trying to highlight.

This would be a very cool feature.
I do not think it is currently possible, please correct me if I am wrong.

**Comment by RobWin** [comment]
Yes, then we might need to enhance this class: https://github.com/resilience4j/resilience4j/blob/master/resilience4j-spring6/src/main/java/io/github/resilience4j/spring6/spelresolver/DefaultSpelResolver.java

**Comment by patpatpat123** [comment]
@LivingLikeKrillin , I tried to come up with a PR, but this is a bit difficult. May I ask if you could help with this? I have couple of test cases ready (fallback methods in some springboot @Service classes) to help test if needed
